File: 560-subarray-sum-equals-k/subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    const map = new Map()
    map.set(0, 1)
    let currentSum = 0
    let count = 0

    for (let num of nums) {
        currentSum += num
        const complement = currentSum - k
        if (map.has(complement)) {
            count += map.get(complement)
        }
        map.set(currentSum, (map.get(currentSum) || 0) + 1)
    }
    return count
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_sums = defaultdict(int)
        prefix_sums[0] = 1
        current_sum = 0
        count = 0

        for num in nums:
            current_sum += num
            difference = current_sum - k
            count += prefix_sums[difference]
            prefix_sums[current_sum] += 1

        return count